Add overheating/freezing mechanics to RD servers and Telecomms

timothymtorres opened the pull request:
## About The Pull Request
_Did you know that the RD server and Telecomms areas had coldrooms but were entirely cosmetic from a gameplay mechanic?_

### Not anymore!!! Inspired by #20878

![dreamseeker_oH7vtbSSzV](https://github.com/user-attachments/assets/4fe391aa-9a35-4432-a244-9af2d9202c76)

RD Servers now:
- Spawn with a thermomachine (if not already present)
- Generates heat while active
- Turns off when overheating or freezing
- Operates between 140K to 220K

Telecomms equipment now:
- Spawns with a thermomachine and heat exchange pipes 
- Generates heat when triggered (speaking = heat)
- Turns off when overheating or freeezing
- Uses power when triggered for all equipment (for consistency)
- Operates between 40K to 120K

What happens when a piece of equipment overheats or freezes? It just shuts off and waits until it reaches its normal temperature operating range. When machines are upgraded with higher tier parts it makes them generate less heat.

##### _(note - space is too cold of an environment for equipment to function)_

## Why It's Good For The Game
This equipment was missing atmos interactions despite obviously being designed for cooling environments. Now it gives the game a little more depth to already existing subsystems. It also gives engineering more projects to upgrade and work on during their downtime.

## Changelog
:cl:
add: Add overheating/freezing mechanics to RD servers and Telecomms equipment. Both will now spawn with their own thermomachine and atmos pipes to regulate temperature. The more you talk on the radio or message via PDA, the more heat that is generated. Tcomms/RD equipment off the station level bypass atmos requirements to prevent problems.
balance: Increase the power requirements for RD servers and Telecomms equipment.
add: Add extra atmos air alarms in Telecomms/RD server areas that monitor the internal temperature of the room and trigger hazard alerts when temperature is outside operation limits for machinery.
add: Telecomms equipment and RD servers can now be upgraded with higher tier parts that makes them generate less heat.
code: Move fluff tcomms equipment to ruin code
del: Remove deprecated birdstation tcomms equipment
/:cl:

timothymtorres committed:
Add heating mechanics to RD servers and Telecomms

MrMelbert commented:
This was removed a while back for causing atmos churn (or in other words, lag) 

We should probably test merge it to see if it's still bad on that

Metekillot commented:
It's already pretty common for the supermatter to become permanently indestructible due to atmos lag. Are we going to end up with permanently broken comms and R&D due to this same thing?

timothymtorres committed:
Fix tcomm and RD areas for stations

MoonDeer728 commented:
> can you make it give a warning on engineering comms that telecomms is too cold/hot? Tcomms is already foreign grounds, and it'd be easier to fix.

no

OrionTheFox commented:
Just to make sure, does this affect /obj/machinery/telecomms/relay (and its subtypes)? 
These is used for multiz/ruins/syndies and I don't think any of them are going to be very compatible with getting a whole cooling setup

Pickle-Coding commented on code/game/machinery/_machinery.dm:
```diff
@@ -301,6 +311,28 @@
+		return FALSE
+
+	var/datum/gas_mixture/enviroment = local_turf.return_air()
+	enviroment.temperature += amount / enviroment.heat_capacity()
```
You should add some sort of checks to make sure it won't get to ridiculously high temperatures from excessively low heat capacity. Could make it thermal conduction based with an imaginary object at a given temperature and heat capacity, or you could have the servers itself contribute its own heat capacity to the calculation.

Pickle-Coding commented:
I would recommend having it process through atmos processing to release heat. It can accumulate heat from people talking like normal, but it only releases that heat during an atmos tick. It should also scale the heat it outputs in the air to the ratio of the atmos subsystem tick delay and the actual real time it took atmos to tick. This should make the effectiveness of cooling telecommunications be independant of atmos lag.

Likteer commented:
Could we turn random tcomm failure events into something fixable, based around the heat system?

timothymtorres committed:
Remove duplicate tgs server (in wall) from bus ruin

timothymtorres committed:
Remove deprecated birdstation tcomm objects

timothymtorres committed:
Move adminbus fluff tcomms to ruin file

timothymtorres committed:
Move fluff tcomms ruin objects to own file

timothymtorres committed:
Add update paths script

timothymtorres committed:
Add temperature immunity to tcomms equipment off station

timothymtorres commented:
> can you make it give a warning on engineering comms that telecomms is too cold/hot? Tcomms is already foreign grounds, and it'd be easier to fix.

If it overheats/freezes then how would the warning message be sent over the radio? It's a catch-22. A smarter fix would be to update the temperature settings for the air alarms linked to each area and that would trigger an atmos alert.

> It's already pretty common for the supermatter to become permanently indestructible due to atmos lag. Are we going to end up with permanently broken comms and R&D due to this same thing?

The complexity of the SM gas/heat interactions isn't even comparable. This only affects heat, no other gas reactions are happening.

> Just to make sure, does this affect /obj/machinery/telecomms/relay (and its subtypes)? These is used for multiz/ruins/syndies and I don't think any of them are going to be very compatible with getting a whole cooling setup

All tcomms equipment off the station has a special variable to bypass atmos interactions so it should not cause any problems.

> I would recommend having it process through atmos processing to release heat. It can accumulate heat from people talking like normal, but it only releases that heat during an atmos tick. It should also scale the heat it outputs in the air to the ratio of the atmos subsystem tick delay and the actual real time it took atmos to tick. This should make the effectiveness of cooling telecommunications be independant of atmos lag.

All decent suggestions. Let me see what I can do.

> Could we turn random tcomm failure events into something fixable, based around the heat system?

You could spawn a bunch of heat that should short-term disable comms. However I don't see this being reasonable for something like an Ion storm which is an electromagnetic event.

timothymtorres committed:
Add temperature ignore settings to RD servers off station

moocowswag commented:
> If it overheats/freezes then how would the warning message be sent over the radio? It's a catch-22. 

Thats like saying we cant have the sm give us a warning about it delamming because it already blew up. Warnings are meant to be given out before the bad thing happens, air alarms are good enough to give the warning but you would think a machine who's only usage is to receive sort and send radio comms would be capable of sending a pre-set recording to the appropriate comms when it detects its getting close to nonfunctional temperatures, just like the automated announcer that speaks whenever someone joins. 

It could be added to the automated announcer as a compromise.

timothymtorres committed:
Change tcomm & rd server areas

timothymtorres committed:
Add air alarm mapping helpers for RD server and Tcomms

timothymtorres committed:
Add RD server and Tcomms temperature defines

timothymtorres committed:
Add Tcomms and RD server gas sensors

timothymtorres committed:
Add Tcomms & RD server room tlv datums

timothymtorres committed:
Fix air alarms not updating icon

timothymtorres committed:
Fix air alarms to use tlv pressure & temp settings

timothymtorres committed:
Fix air alarm helper to ignore gas when appropriate

timothymtorres committed:
Add air alarm tlv for tcomms and rd server

timothymtorres committed:
Fix RD server & Tcomm areas to have tcomm tiles

timothymtorres committed:
Fix duplicate pipe

timothymtorres committed:
Fix indenting

MrMelbert commented:
I will say I don't like the Icebox mapping changes, the RND servers being cooled by the outside air is a very cool flavor element which is being axed

MrMelbert commented:
Also why did we scrunch up a lot of the RND rooms like on Meta? Wasn't it big enough already?

![image](https://github.com/user-attachments/assets/6b22435a-91b3-424a-af6c-3177fff0a1f8)

timothymtorres commented:
> I will say I don't like the Icebox mapping changes, the RND servers being cooled by the outside air is a very cool flavor element which is being axed

I liked it too but it was a necessary evil since Icebox atmos for the outside air is 180K and the rest of tcomms/RD areas are 80K. 

To support the temperature at 80K:

- Remove flooring to Icebox RD server and use a thermomachine

To support the temperature at 180K:

- Change the tcomms/rd server turf settings for a large list of tiles and their subtypes (maybe over a dozen)
- Map in a custom thermomachine at 180K for every map
- Change the temperature settings for all tcomms and RD server equipment

The other thing is right now you can use the space cooling loops to keep the equipment cool without it freezing. (since there is a large thermal penalty from using heat pipes) Changing the equipment to 180K would make this not possible without using temperature gates and pumps.

---

> Also why did we scrunch up a lot of the RND rooms like on Meta? Wasn't it big enough already?

Since a valid complaint is atmos lag, I went and decreased the amount of active gas turfs to boost performance.

MrMelbert commented:
> I liked it too but it was a necessary evil since Icebox atmos for the outside air is 180K and the rest of tcomms/RD areas are designed to be 80K.

Not possible to tweak the temperature range to allow both, then? 

Or what if we had some unique turfs for icebox, like an ice geyser (which spits colder-than-normal air) underneath the servers?

> Since a valid complaint is atmos lag, I went and decreased the amount of active gas turfs to boost performance.

I would suggest not remapping all the rooms to be smaller until that is quantified tbh, I don't think the smaller looks better (map design wise) so it'd be nice to keep them unless otherwise necessary

timothymtorres committed:
Change RD server room to use ice temps

timothymtorres committed:
Merge branch 'master' into rd-server-generate-heat

timothymtorres committed:
Fix messed up merge conflict

timothymtorres committed:
Change all maps except icebox to accomidate RD server temp changes

timothymtorres committed:
Change temp thresholds to be +- 20

timothymtorres committed:
Fix define linter error

timothymtorres committed:
Revert delta RD server window

timothymtorres committed:
Fix icebox RD server room to be old style

timothymtorres committed:
Change telecomms machinery to only generate heat during atmos tick

timothymtorres committed:
Move Meta Tcomm air alarm to corner of room

timothymtorres committed:
Fix air alarm ignoring disabled settings

timothymtorres committed:
Fix air alarm sensor turf signals

timothymtorres committed:
Fix air alarm sensor signals

timothymtorres committed:
Add tcomms gas monitor to global gas chamber list

timothymtorres committed:
Merge branch 'master' into rd-server-generate-heat

timothymtorres committed:
Add comments for local turf checks

timothymtorres committed:
Fix sound loop not initializing before used in update_power()

timothymtorres committed:
Change span for machinery examine to be warning

timothymtorres committed:
Fix pipe issues

timothymtorres committed:
Change RD server pipes to be visible

timothymtorres committed:
Remove double thermomachine from tcomms

timothymtorres committed:
Add atmos machine defines

timothymtorres committed:
Refactor machines to use bitflags, defines, and atmos processing

timothymtorres committed:
Give heat capacity define to thermomachine

timothymtorres committed:
Add ntnet relays to atmos sensitive

timothymtorres commented:
> Not possible to tweak the temperature range to allow both, then?

Alright so I set the temperature range for RD servers to 180K, converted all maps to use 180K temp turfs for RD server areas, and mapped in 180K thermomachines. It was a lot of extra work but it looks good now.
 
> I would suggest not remapping all the rooms to be smaller until that is quantified tbh, I don't think the smaller looks better (map design wise) so it'd be nice to keep them unless otherwise necessary

Reverted the smaller designs as requested.

The code is finally in a test mergeable state and ready for a final review.

timothymtorres committed:
Merge branch 'master' into rd-server-generate-heat

timothymtorres committed:
Merge branch 'master' into rd-server-generate-heat

timothymtorres committed:
Fix missing pipe for wawa TCOMMs

timothymtorres committed:
Merge branch 'rd-server-generate-heat' of https://github.com/timothymtorres/tgstation into rd-server-generate-heat

timothymtorres committed:
Fix tram TCOMM power issue

timothymtorres committed:
Rebalance TCOMM messages heat due to machinery radio spam

timothymtorres committed:
Adjust temperature warning settings for air alarms

timothymtorres committed:
Tier parts upgrade for Tcomm/RD servers generate less heat

timothymtorres committed:
Fix tram Telecomms wire

Autisem commented:
on round 240250 there was some pretty bad lag which causes tcoms to be down for pretty much the entire round due to overheating on default setup. took me opening cooled nitrogen canister in the room to resolve(which was also slow since atmos ticked every like 3 sconds)

Autisem commented:
on round 240252 lag was so bad it killed tcoms for practicly the entire round, even flooding it with cold gas did not work as the atmos ticked so slowly it didnt spread properly

Jacquerel commented:
It seems like this makes a reliable system into an unreliable one with very little justification in the "why this is good for the game" section of the PR other than "it adds depth".
I am unconvinced that anyone actually _wants_ this important game system to have "depth" which comes at the expense of turning something which works into something which often will not and will break for reasons most people in the round don't understand and haven't interacted with, at the expense of their ability to talk and interact with other players when "talking to people" is virtually our most important core game system. 
I think this is probably not a good idea.

Timberpoes commented:
Wise words from Jacquerel phrased better than I could. I concur.

timothymtorres committed:
Put soft cap on max amount of heat generated per atmos tick

timothymtorres committed:
account for spam

timothymtorres commented:
Finally have free time to address this.

> It seems like this makes a reliable system into an unreliable one 

It changes a _closed-loop_ system that nobody interacts with to a _open-sandbox system_ that is more affected by game mechanics.

> with very little justification in the "why this is good for the game" section of the PR other than "it adds depth". 

What is depth? Does it always have to be beneficial?

<details>
<summary>Why did we go from this:</summary>

![600px-Singularity_engine](https://github.com/user-attachments/assets/ed481f18-4e95-4a6e-8f35-153145a43253)

</details>

<details>
<summary>To this:</summary>

![e7141f10-fbaf-11e6-97c3-b1d19b3da586](https://github.com/user-attachments/assets/3bc5bad5-d41c-4541-8605-ce2f4fbf486a)

</details>

When it comes to performance, stability, safety, setup, and power generation the SM is the worse choice:
- It relies _heavily_ on atmos interactions which consumes more processing ticks than the traditional singularity engine
- It blows up at a higher frequency than the singularity engine. (and the anomalies force a shuttle call so people can't ignore it)
- There is a required knowledge ceiling for gas interactions and pipes
- The setup requires many more steps with little room for error before it delams
- Power generation depends on a complicated formula of gases, temperature, and pressure
- Trying to do anything fancy with gases requires upgrades to improve it

VS

The singularity engine where engineers activate emitters, activate shields, activate the radiation emitter and then boom! Guaranteed stable power generation for the rest of the round without _any_ added atmos lag!

Why am I using this example? 

Because we traded a _closed-loop system_ for an _open-sandbox system_ in regards to power generation. _That's depth!!!_ By every metric the SM is a worse choice but since it has depth it lets people be creative. [(Losing is fun!)](https://dwarffortresswiki.org/images/thumb/9/91/Fun_kruggsmash.jpg/596px-Fun_kruggsmash.jpg) I think the purpose of this PR is nearly identical to the purpose of the SM change so I'm using it as a strong reference. Especially since both boil down to adding mechanics that generates heat and requires cooling.

> I am unconvinced that anyone actually _wants_ this important game system to have "depth" 

![chrome_wt5mVY9TBg](https://github.com/user-attachments/assets/316520a6-fda4-4153-93e4-79e772dc3d71)

17 people want, 10 people no want

@LemonInTheDark and @MrMelbert have both been consulted on code implementation and mapping changes so they are welcome to chime in on their thoughts.

> which comes at the expense of turning something which works into something which often will not and will break for reasons most people in the round don't understand and haven't interacted with, at the expense of their ability to talk and interact with other players when "talking to people" is virtually our most important core game system.

I should clarify that when it overheats, it shuts off (aka clogging) while it cools to the correct temperature threshold. So it flip flops between online and offline depending on radio traffic. Not every machine generates heat simultaneously. One section of the room might overheat resulting in comms to be shutoff for specific departments or the processor/buses which affect how fast a message is sent and scrambled messages. It's kinda cool seeing this happen because you don't normally see any of those machines interact with telecomms separately since it's usually just the APC that gets flipped to the on/off switch during sabotage attempts.

---

Anyway the crux of this discussion - _should telecomms radio messages be an infinite resource on the station?_ 

I don't think it should.

Every department starts with limited supplies and resources. Medbay only has two stasis beds and a few spare medkits. Engineering only has a small chunk of materials, a few pairs of insulated gloves, etc. Mining only has a few miner sets. I could go on but you get the idea. This gives reasons for the players to build, create, improve, or improvise solutions. It's why our game is a sandbox.

People now have a reason to use higher tier parts to upgrade telecomms equipment (less heat) or to modify the cooling loop with improvements to make it more resilient. (heat exchange pipes to space, more thermomachines, better gas with heat capacity, etc.) It also adds new opportunities for people to sabotage the equipment without getting caught. If you leak the coolant or fuckup the thermomachine you can time a TCOMM blackout.

> Testmerging has demonstrated that this in the modern day still _doesn't_ play nicely with our server atmos tick, because atmos is our least prioritised system and so is wildly unreliable in terms of when effects incurred by people using it will happen.

I was present when the TM went live and watched players on low to mid pop rounds for hours. Engineers were tinkering with it and I observed several of them improve the cooling loops. I do admit I fucked up two maps (Wawa & Tram) piping/wiring systems since I had to redo all the maps multiple times during the review process but it was a quick single pipe/wire fix on one turf. 

I did not balance my system around +100 server population which is where they started to have severe problems. To be fair, most departments experience severe supply bottlenecks at these numbers as well. There was a ton of radio spam that was generated from machines that I did not account for when I came up with my heat equation. (research completed, medibot healing, beepsky, cargo deliveries, cryotube patient ejected, etc.)

The TM was reverted before I could apply balance tweaks which was frustrating. I said on the discord that it felt like a knee-jerk reaction because you need to stress test PRs like this but I was denied the opportunity. 

> Overall I think this is probably not a good idea worth pursuing. Even fixing the atmos issues (if this is even possible), the result would be that we end up with a worse and less fun game than we had before we started. I do not think it would be worth your time to attempt, but I can't stop you.

TG's motto has always been "_Improve don't remove_". As Melbert pointed out earlier this heating system already existed in the past. I actually dug around and found the PR where it was mentioned,

- #25132

>> This tweak slightly extends Omegastations telecoms room downwards to make room for the new in built freezer, this should allow the telecoms room on Omegatation to be more consistent with the other maps even though telecoms equipment no longer heats up this still is something that is on all other maps so there is not reason it should not also be installed on this map this also i guess adds the illusion of the telecoms equipment requiring cooling down even though it is not longer strictly necessary for the operation of the equipment.

Atmos was not optimized at the time so it was removed. (LINDA has undergone tremendous improvements since then)

Regardless of this PR, the atmos processing is already happening on most maps for Tcomms and RD servers. The cooling loops and thermomachines are mapped in that chew up atmos with no gameplay functionality.

---

tldr; I think telecomms should not be an infinite resource and require an atmos sandbox system as a constraint

EuclidianGeometer commented:
> Anyway the crux of this discussion - _should telecomms radio messages be an infinite resource on the station?_
> 
> I don't think it should.
> 
> Every department starts with limited supplies and resources. Medbay only has two stasis beds and a few spare medkits. Engineering only has a small chunk of materials, a few pairs of insulated gloves, etc. Mining only has a few miner sets. I could go on but you get the idea. This gives reasons for the players to build, create, improve, or improvise solutions. It's why our game is a sandbox.
> 
> People now have a reason to use higher tier parts to upgrade telecomms equipment (less heat) or to modify the cooling loop with improvements to make it more resilient. (heat exchange pipes to space, more thermomachines, better gas with heat capacity, etc.) It also adds new opportunities for people to sabotage the equipment without getting caught. If you leak the coolant or fuckup the thermomachine you can time a TCOMM blackout.


This doesn't in any way make an argument for why telecommunications messages being limited is good. I think it's good for it to be sabotage, but the radio is one of the core ways players interact with the game- the dang roleplaying in the roleplaying game. It shouldn't be something that stops working by itself without someone going out of their way to sabotage it as an antagonist, or a big dramatic event like a rod or a meteor or other interference that creates a story, in my opinion. It's fun to talk on the radio! I like to listen to people and bicker.

Github users are not the most reliable measure of if an idea is good or not- the upvotes in of themselves are a data point, but not a super meaningful one. They tend towards coders, which means they tend towards people who like to tinker with a system. More complexity, the idea of the thing that has more you can do with it, is disporportionately compelling to that sort of person. That's not a diss on anyone, it's just the way it is! 


Your comparison to the SM doesn't quite fit, because the basic roundstart setup more or less works at default at any population. (And the depth of the SM doesn't have much practical purpose at the moment, since there's not much to do with extra power and people don't find being out of power that fun for the most part. I remember at one point people were tinkering with some ideas for incentives to generate extra power, but they were along the lines of bonuses to other systems, not just making things basically function.)

ElGitificador commented:
![imagen](https://github.com/user-attachments/assets/aa533369-934d-4c6c-ba03-084c5136fa3a)

> I am unconvinced that anyone actually wants this important game system to have "depth"

Good thing mantainers care for players opinions on tg

Hotmud9 commented:
this does beg the question of "why the hell are they oxygenless in the first place"
